File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/logic/function/process/Process.java
package io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process;

import io.github.arcaneplugins.levelledmobs.bukkit.LevelledMobs;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.LmFunction;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.Action;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.condition.Condition;
import io.github.arcaneplugins.levelledmobs.bukkit.logic.preset.Preset;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.function.Supplier;
import org.bukkit.scheduler.BukkitRunnable;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.configurate.CommentedConfigurationNode;

@SuppressWarnings("unused")
public class Process {

    /*
    TODO
        - isEnabled
        - factor in isEnabled
     */

    /* vars */

    private final String identifier;
    private final String description;
    private final CommentedConfigurationNode node;
    private final LmFunction parentFunction;
    private final long delay;
    private final Set<Preset> presets = new LinkedHashSet<>();
    private final List<Action> actions = new LinkedList<>();
    private final List<Condition> conditions = new LinkedList<>();
    private boolean exit = false;

    /* constructors */

    public Process(
        final @NotNull String identifier,
        final @NotNull String description,
        final @NotNull CommentedConfigurationNode node,
        final @NotNull LmFunction parentFunction,
        final long delay
    ) {
        this.identifier = identifier;
        this.description = description;
        this.node = node;
        this.parentFunction = parentFunction;
        this.delay = delay;
    }

    /* methods */

    public void call(
        final @NotNull Context context,
        final boolean overrideConditions
    ) {
        final Supplier<Boolean> canRun = () -> overrideConditions || conditionsApply(context);

        if(getDelay() == 0) {
            if(!canRun.get()) return;
            runActions(context);
        } else {
            new BukkitRunnable() {
                @Override
                public void run() {
                    if(!canRun.get()) return;
                    runActions(context);
                }
            }.runTaskLater(LevelledMobs.getInstance(), getDelay());
        }
    }

    private boolean conditionsApply(final @NotNull Context context) {
        final int totalConditions = getConditions().size();

        if(totalConditions == 0)
            return true;

        final float conditionsPercentageRequired = 1.0f; //TODO configurable
        int conditionsMet = 0;

        for(var condition : getConditions()) {
            if (condition.applies(context)) {
                conditionsMet++;
            }
        }

        return (conditionsMet * 1.0f / totalConditions) >= conditionsPercentageRequired;
    }

    private void runActions(final @NotNull Context context) {
        exit = false;
        for(var action : getActions()) {
            if(exiting()) return;
            action.run(context);
        }
        exit = false;
    }

    /* getters and setters */

    @NotNull
    public String getIdentifier() { return identifier; }

    @NotNull
    public String getDescription() { return description; }

    @NotNull
    public Set<Preset> getPresets() { return presets; }

    @NotNull
    public CommentedConfigurationNode getNode() { return node; }

    @NotNull
    public List<Action> getActions() { return actions; }

    @NotNull
    public List<Condition> getConditions() { return conditions; }

    @NotNull
    public LmFunction getParentFunction() { return parentFunction; }

    public long getDelay() { return delay; }

    public boolean exiting() {
        return exit;
    }

    public void setExiting(final boolean exit) {
        this.exit = exit;
    }

}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/logic/function/process/Process.kt
package io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process

import io.github.arcaneplugins.levelledmobs.bukkit.LevelledMobs
import io.github.arcaneplugins.levelledmobs.bukkit.debug.DebugCategory
import io.github.arcaneplugins.levelledmobs.bukkit.logic.context.Context
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.LmFunction
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.action.Action
import io.github.arcaneplugins.levelledmobs.bukkit.logic.function.process.condition.Condition
import io.github.arcaneplugins.levelledmobs.bukkit.logic.preset.Preset
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log
import io.github.arcaneplugins.levelledmobs.bukkit.util.Log.debug
import org.bukkit.scheduler.BukkitRunnable
import org.spongepowered.configurate.CommentedConfigurationNode
import java.util.function.Supplier

class Process(
    val identifier: String,
    val description: String,
    val node: CommentedConfigurationNode,
    val parentFunction: LmFunction,
    val delay: Long
) {
    val conditions: MutableList<Condition> = mutableListOf()
    val actions: MutableList<Action> = mutableListOf()
    val presets: MutableSet<Preset> = mutableSetOf()
    var exit = false

    fun call(
        context: Context,
        overrideConditions: Boolean
    ){
        debug(DebugCategory.FUNCTIONS_GENERIC) {"Calling process ${identifier}."}

        //val canRun:Supplier<Boolean> = { overrideConditions || conditionsApply(context) }
        val canRun = Supplier {
            overrideConditions || conditionsApply(context)
        }

        if (delay == 0L){
            if (!canRun.get()) {
                debug(DebugCategory.FUNCTIONS_GENERIC) {"Process $identifier did not meet canRun requirements."}
                return
            }
            runActions(context)
        }
        else{
            debug(DebugCategory.FUNCTIONS_GENERIC) {"Process $identifier has a delay of $delay ticks, running later."}
            object : BukkitRunnable() {
                override fun run() {
                    debug(DebugCategory.FUNCTIONS_GENERIC) {"Process $identifier is now running after its delay of $delay ticks."}
                    if (!canRun.get()) {
                        debug(DebugCategory.FUNCTIONS_GENERIC) {"Process $identifier did not meet canRun requirements."}
                        return
                    }
                    runActions(context)
                }
            }.runTaskLater(LevelledMobs.lmInstance, delay)
        }
    }

    private fun conditionsApply(context: Context): Boolean{
        val totalConditions: Int = conditions.size

        if (totalConditions == 0) return true

        val conditionsPercentageRequired = 1.0f //TODO configurable
        var conditionsMet = 0

        for (condition in conditions){
            val result = condition.applies(context)
            debug(DebugCategory.CONDITION){ "Condition: ${condition.name}, entity: ${context.entity}, result: $result" }
            if (result){
                conditionsMet++
            }
        }

        return (conditionsMet * 1.0f / totalConditions) >= conditionsPercentageRequired
    }

    private fun runActions(context: Context){
        debug(DebugCategory.FUNCTIONS_GENERIC) {"Process $identifier is running its actions."}
        exit = false
        for (action in actions){
            if (exit) {
                debug(DebugCategory.FUNCTIONS_GENERIC) {"Process $identifier is exiting early (exit=${exit})."}
                return
            }
            action.run(context)
        }
        exit = false
    }


}